Automattic/wp-calypso
# Replace react-stripe-elements with react-stripe-js

#### Changes proposed in this Pull Request

We currently use the package `react-stripe-elements` in Calypso for checkout and payment method management. However, that package is outdated and doesn't support React 17. In order for the rest of calypso to update (see #54793), we'll need to migrate to `@stripe/react-stripe-js`.

This PR follows the migration guide at https://github.com/stripe/react-stripe-js/blob/master/docs/migrating.md, and attempts to retain as much of our existing API as possible by modifying the `@automattic/calypso-stripe` package, which is the interface used by calypso to interact with Stripe. 

Unfortunately, there are some API changes that cannot be easily hidden by `@automattic/calypso-stripe`. Notably, Stripe Elements (the React components that hold the iframed credit card input fields) are no longer automatically accessed by the Stripe API functions; this means that we have to pass a reference to the card number element to the functions directly. There's a convenient `useElements` React hook to do that, but it has to be done inside a React component, so this has to be done at the level of the consumer. The good news is that the only place this is required is when adding a new card, so the impact is limited.

Fixes https://github.com/Automattic/wp-calypso/issues/55944

#### Testing instructions

We'll need to try various instances of Stripe-powered card processing in calypso, making sure to also test the failure cases of each of the situations described.

If you have the API sandboxed, you should be able to use [Stripe test cards](https://stripe.com/docs/testing) for the below tests, which includes cards that can fail and cards that trigger an authentication dialog (3DS). Some examples you can use:

- Success, no authentication required: `4242 4242 4242 4242`
- Success, requires authentication: `4000 0027 6000 3184`
- Failure, no authentication required: `4000 0000 0000 9995`

Test cases:

- Adding a new card in the payment methods list
  - Visit `/me/purchases/add-payment-method`
  - With 3DS
  - Without 3DS
- Adding a new card in the payment method assignment form
  - Visit `/me/purchases` and click on an active subscription, then click "Add Payment Method" or "Change Payment Method".
  - With 3DS
  - Without 3DS
- Making a purchase with a new card in checkout
  - With 3DS
  - Without 3DS
- Making a purchase in checkout with a stored card
  - With 3DS
  - Without 3DS
- Making a purchase in the one-click upsell
  - Visit `/checkout/offer-quickstart-session/1234/[YOUR SITE HERE]` when you own a personal/premium plan.
  - With 3DS
  - Without 3DS
- Making a purchase in checkout with Stripe Sources (iDEAL is a good one)
  - Account currency must be EUR and apply D45029-code to your sandbox to force iDEAL to be available.
- Making a purchase in checkout with Google Pay
  - Apply D60830-code to allow using the REST API through a tunnel.
  - Use localtunnel to visit localhost over https or use the calypso.live URL.
  - You must have an existing card in your Google wallet.
- Making a purchase in checkout with Apple Pay
  - Apply D60830-code to allow using the REST API through a tunnel.
  - Disable the proxy.
  - Enable global http access to your sandbox.
  - You must have an existing card in your Apple Pay wallet on your local Safari instance.
  - Use Mac OS Safari.
  - Use localtunnel to visit localhost over https at the subdomain `test-calypso`.
